nucypher/policy/kits.py

```diff
@@ -0,0 +1,250 @@
+"""
+This file is part of nucypher.
+
+nucypher is free software: you can redistribute it and/or modify
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+nucypher is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+
+from typing import Dict, NamedTuple, Optional, Tuple, List, Iterable
+
+from bytestring_splitter import BytestringSplitter, BytestringKwargifier, VariableLengthBytestring
+from constant_sorrow.constants import (
+    NOT_SIGNED,
+    UNKNOWN_SENDER,
+    DO_NOT_SIGN,
+    SIGNATURE_TO_FOLLOW,
+    SIGNATURE_IS_ON_CIPHERTEXT,
+    NOT_SIGNED,
+    )
+from eth_typing import ChecksumAddress
+from eth_utils import to_checksum_address, to_canonical_address
+
+from nucypher.crypto.splitters import capsule_splitter, key_splitter, signature_splitter, checksum_address_splitter
+import nucypher.crypto.umbral_adapter as umbral # need it to mock `umbral.encrypt`
+from nucypher.crypto.umbral_adapter import PublicKey, VerifiedCapsuleFrag, Capsule, Signature
+
+
+class MessageKit:
+    """
+    All the components needed to transmit and verify an encrypted message.
+    """
+
+    @classmethod
+    def author(cls,
+               recipient_key: PublicKey,
+               plaintext: bytes,
+               signer: 'SignatureStamp',
+               sign_plaintext: bool = True
+               ) -> Tuple['MessageKit', Signature]:
+        if signer is not DO_NOT_SIGN:
+            # The caller didn't expressly tell us not to sign; we'll sign.
+            if sign_plaintext:
+                # Sign first, encrypt second.
+                sig_header = SIGNATURE_TO_FOLLOW
+                signature = signer(plaintext)
+                capsule, ciphertext = umbral.encrypt(recipient_key, sig_header + bytes(signature) + plaintext)
+                signature_in_kit = None
+            else:
+                # Encrypt first, sign second.
+                sig_header = SIGNATURE_IS_ON_CIPHERTEXT
+                capsule, ciphertext = umbral.encrypt(recipient_key, sig_header + plaintext)
+                signature = signer(ciphertext)
+                signature_in_kit = signature
+            message_kit = cls(ciphertext=ciphertext,
+                              capsule=capsule,
+                              sender_verifying_key=signer.as_umbral_pubkey(),
+                              signature=signature_in_kit)
+        else:
+            # Don't sign.
+            signature = sig_header = NOT_SIGNED
+            capsule, ciphertext = umbral.encrypt(recipient_pubkey_enc, sig_header + plaintext)
```

Comment:
Doesn't seem like this line gets tested - ideally a test should have failed.

```suggestion
            capsule, ciphertext = umbral.encrypt(recipient_key, sig_header + plaintext)
```